File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/services/AIServiceCollection.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.services;

import com.microsoft.semantickernel.Verify;
import com.microsoft.semantickernel.ai.embeddings.EmbeddingGeneration;
import com.microsoft.semantickernel.ai.embeddings.TextEmbeddingGeneration;
import com.microsoft.semantickernel.chatcompletion.ChatCompletion;
import com.microsoft.semantickernel.textcompletion.TextCompletion;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import javax.annotation.Nullable;

/** A collection that can hold AI services */
public class AIServiceCollection {
    // A constant key for the default service
    private static final String DefaultKey = "__DEFAULT__";

    private static final List<Class<? extends AIService>> KNOWN_SERVICES =
            Arrays.asList(
                    EmbeddingGeneration.class,
                    TextEmbeddingGeneration.class,
                    TextCompletion.class,
                    ChatCompletion.class);

    // A dictionary that maps a service type to a nested dictionary of names and service instances
    // or factories
    // private readonly Dictionary<Type, Dictionary<string, object>> _services = new();
    private final Map<Class<? extends AIService>, Map<String, Supplier<? extends AIService>>>
            services = new HashMap<>();

    // A dictionary that maps a service type to the name of the default service
    private final Map<Class<? extends AIService>, String> defaultIds = new HashMap<>();

    /**
     * Registers a singleton service instance with the default name.
     *
     * @param service The service instance.
     * @param serviceType The type of the service.
     */
    public <T extends AIService> void setService(T service, Class<T> serviceType) {
        setService(DefaultKey, (Supplier<T>) () -> service, true, serviceType);
    }

    /**
     * Registers a singleton service instance with an optional name and default flag.
     *
     * @param name The name of the service, or null for the default service.
     * @param service The service instance.
     * @param setAsDefault Whether the service should be the default for its type.
     * @param serviceType The type of the service.
     */
    public <T extends AIService> void setService(
            String name, T service, boolean setAsDefault, Class<T> serviceType) {
        setService(name, (Supplier<T>) () -> service, setAsDefault, serviceType);
    }

    /**
     * Registers a transient service factory with the default name.
     *
     * @param factory The factory function to create the service instance.
     * @param serviceType The type of the service.
     */
    public <T extends AIService> void setService(Supplier<T> factory, Class<T> serviceType) {
        setService(DefaultKey, factory, true, serviceType);
    }

    /**
     * Registers a transient service factory with an optional name and default flag.
     *
     * @param name The name of the service, or null for the default service.
     * @param factory The factory function to create the service instance.
     * @param setAsDefault Whether the service should be the default for its type.
     * @param specificServiceType The type of the service.
     */
    public <T extends AIService> void setService(
            @Nullable String name,
            Supplier<T> factory,
            boolean setAsDefault,
            Class<T> specificServiceType) {

        // Validate the factory function
        if (factory == null) {
            throw new IllegalArgumentException();
        }

        // if this is an implementation of a known service type, register it as such
        getKnownServiceTypes(specificServiceType)
                .forEach(
                        serviceType -> {
                            // Get or create the nested dictionary for the service type
                            Map<String, Supplier<? extends AIService>> namedServices =
                                    this.services.computeIfAbsent(
                                            serviceType, (ignore) -> new HashMap<>());

                            // Set as the default if the name is empty, or the default flag is true,
                            // or there is no default name for the service type.
                            if (name == null || setAsDefault || !this.hasDefault(serviceType)) {
                                // Update the default name for the service type
                                this.defaultIds.put(serviceType, name == null ? DefaultKey : name);
                            }

                            // Register the factory with the given name
                            namedServices.put(name == null ? DefaultKey : name, factory);
                        });
    }

    private <T extends AIService> List<Class<? extends AIService>> getKnownServiceTypes(
            Class<T> serviceType) {
        ArrayList<Class<? extends AIService>> knownServices =
                KNOWN_SERVICES.stream()
                        .filter(knownServiceType -> knownServiceType.isAssignableFrom(serviceType))
                        .collect(Collectors.toCollection(ArrayList::new));

        knownServices.add(serviceType);
        return knownServices;
    }

    /**
     * Builds an AIServiceProvider from the registered services and default names.
     *
     * @return The AIServiceProvider.
     */
    public AIServiceProvider build() {
        // Create a clone of the services and defaults Dictionaries to prevent further changes
        // by the services provider.
        Map<Class<? extends AIService>, Map<String, Supplier<? extends AIService>>> servicesClone =
                cloneServices(services);

        Map defaultsClone = new HashMap<>(defaultIds);

        return new DefaultAIServiceProvider(servicesClone, defaultsClone);
    }

    /**
     * Creates a clone of the services as a Map.
     *
     * @param services The services Map.
     * @return The clone.
     */
    static <T> Map<Class<? extends T>, Map<String, Supplier<? extends T>>> cloneServices(
            Map<Class<? extends T>, Map<String, Supplier<? extends T>>> services) {
        return Collections.unmodifiableMap(
                services.entrySet().stream()
                        .reduce(
                                new HashMap<>(),
                                (a, b) -> {
                                    a.put(b.getKey(), Collections.unmodifiableMap(b.getValue()));
                                    return a;
                                },
                                (a, b) -> {
                                    a.putAll(b);
                                    return a;
                                }));
    }

    private <T extends AIService> boolean hasDefault(Class<T> type) {
        String defaultName = defaultIds.get(type);
        return !Verify.isNullOrEmpty(defaultName);
    }
}

File: dotnet/src/SemanticKernel.Core/Services/AIServiceCollection.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Linq;

namespace Microsoft.SemanticKernel.Services;

/// <summary>
/// A collection of AI services that can be registered and built into an <see cref="IAIServiceProvider"/>.
/// </summary>
[System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1711:Identifiers should not have incorrect suffix")]
public class AIServiceCollection
{
    // A constant key for the default service
    private const string DefaultKey = "__DEFAULT__";

    // A dictionary that maps a service type to a nested dictionary of names and service instances or factories
    private readonly Dictionary<Type, Dictionary<string, Func<object>>> _services = new();

    // A dictionary that maps a service type to the name of the default service
    private readonly Dictionary<Type, string> _defaultIds = new();

    /// <summary>
    /// Registers a singleton service instance with the default name.
    /// </summary>
    /// <typeparam name="T">The type of the service.</typeparam>
    /// <param name="service">The service instance.</param>
    /// <exception cref="ArgumentNullException">The service instance is null.</exception>
    public void SetService<T>(T service) where T : IAIService
        => this.SetService(DefaultKey, service, true);

    /// <summary>
    /// Registers a singleton service instance with an optional name and default flag.
    /// </summary>
    /// <typeparam name="T">The type of the service.</typeparam>
    /// <param name="name">The name of the service, or null for the default service.</param>
    /// <param name="service">The service instance.</param>
    /// <param name="setAsDefault">Whether the service should be the default for its type.</param>
    /// <exception cref="ArgumentNullException">The service instance is null.</exception>
    /// <exception cref="ArgumentException">The name is empty or whitespace.</exception>
    public void SetService<T>(string? name, T service, bool setAsDefault = false) where T : IAIService
        => this.SetService<T>(name, (() => service), setAsDefault);

    /// <summary>
    /// Registers a transient service factory with the default name.
    /// </summary>
    /// <typeparam name="T">The type of the service.</typeparam>
    /// <param name="factory">The factory function to create the service instance.</param>
    /// <exception cref="ArgumentNullException">The factory function is null.</exception>
    public void SetService<T>(Func<T> factory) where T : IAIService
        => this.SetService<T>(DefaultKey, factory, true);

    /// <summary>
    /// Registers a transient service factory with an optional name and default flag.
    /// </summary>
    /// <typeparam name="T">The type of the service.</typeparam>
    /// <param name="name">The name of the service, or null for the default service.</param>
    /// <param name="factory">The factory function to create the service instance.</param>
    /// <param name="setAsDefault">Whether the service should be the default for its type.</param>
    /// <exception cref="ArgumentNullException">The factory function is null.</exception>
    /// <exception cref="ArgumentException">The name is empty or whitespace.</exception>
    public void SetService<T>(string? name, Func<T> factory, bool setAsDefault = false) where T : IAIService
    {
        // Validate the factory function
        if (factory == null)
        {
            throw new ArgumentNullException(nameof(factory));
        }

        // Get or create the nested dictionary for the service type
        var type = typeof(T);
        if (!this._services.TryGetValue(type, out var namedServices))
        {
            namedServices = new();
            this._services[type] = namedServices;
        }

        // Set as the default if the name is empty, or the default flag is true,
        // or there is no default name for the service type.
        if (name == null || setAsDefault || !this.HasDefault<T>())
        {
            // Update the default name for the service type
            this._defaultIds[type] = name ?? DefaultKey;
        }

        var objectFactory = factory as Func<object>;

        // Register the factory with the given name
        namedServices[name ?? DefaultKey] = objectFactory
                                            ?? throw new InvalidOperationException("Service factory is an invalid format");
    }

    /// <summary>
    /// Builds an <see cref="IAIServiceProvider"/> from the registered services and default names.
    /// </summary>
    /// <returns>An <see cref="IAIServiceProvider"/> containing the registered services.</returns>
    public IAIServiceProvider Build()
    {
        // Create a clone of the services and defaults Dictionaries to prevent further changes
        // by the services provider.
        var servicesClone = this._services.ToDictionary(
            typeCollection => typeCollection.Key,
            typeCollection => typeCollection.Value.ToDictionary(
                service => service.Key,
                service => service.Value));

        var defaultsClone = this._defaultIds.ToDictionary(
            typeDefault => typeDefault.Key,
            typeDefault => typeDefault.Value);

        return new AIServiceProvider(servicesClone, defaultsClone);
    }

    private bool HasDefault<T>() where T : IAIService
        => this._defaultIds.TryGetValue(typeof(T), out var defaultName)
            && !string.IsNullOrEmpty(defaultName);
}
